add gcov and lcov for code coverage metrics for the sdk

lbk003 opened the pull request:
Fixes #13 
Use the modifications in README to see the output of gcov and lcov
The code coverage report can be read from build/out/index.html
This has been tested with the flags -DBUILD_TESTS & -DENABLE_COVERAGE
gcov, lcov and genhtml binaries were renamed : cmake aborted as expected

Commit:
add gcov and lcov for code coverage metrics for the sdk

mickey121 commented on CMakeLists.txt:
```diff
@@ -19,6 +19,12 @@ set(BUILD_TESTS, 0)
+# Code Coverage for sdk
+#####################################
+# To run report on code coverage, add option flag -DBUILD_COVERAGE=1 to cmake command
+set(BUILD_COVERAGE, 0)
```
Does it mean the code coverage is disabled for SDK by default?

vaasu1234 replied:
Yes

vaasu1234 commented on tests/CMakeLists.txt:
```diff
@@ -50,3 +50,43 @@ target_link_libraries(tests ${APPD_SDK_LINK_LIBS} ${CGREEN_LINK_LIBS})
+
+add_custom_command(
+    TARGET run-tests POST_BUILD
+    COMMAND lcov --capture --directory . --output-file appd-iot-cpp-coverage-tmp.info
```
lcov 1.13 version is needed. when installing lcov using brew it installed lcov 1.12 and it was looking for .da file instead of .gcda files.
It worked after I installed it from http://ltp.sourceforge.net/coverage/lcov.php.

lbk003 replied:
Done

vaasu1234 commented on CMakeLists.txt:
```diff
@@ -19,6 +19,12 @@ set(BUILD_TESTS, 0)
+#####################################
+# To run report on code coverage, add option flag -DBUILD_COVERAGE=1 to cmake command
+set(BUILD_COVERAGE, 0)
+
```
I would include TESTS along with coverage.

if(BUILD_COVERAGE)
  set(BUILD_TESTS 1)

So, that both flags are independent of each other.

lbk003 replied:
Done

vaasu1234 commented on README.md:
````diff
@@ -90,6 +90,11 @@ Only sdk and sample targets are built by default. To build tests, run
 $ cmake .. -DBUILD_TESTS=1
 $ make
 ```
+If you want to view the code coverage done by the tests, set the flag -DBUILD_COVERAGE in addition to the -DBUILD_TESTS flag
````
Add the tools that coverage is dependent on (lcov, gcov, genhtml)

LCOV version 1.13-16-ge675080

lbk003 replied:
Done. (Lcov Version comment is dup of the comment above.)

vaasu1234 commented on CMakeLists.txt:
```diff
@@ -19,6 +19,12 @@ set(BUILD_TESTS, 0)
+# Code Coverage for sdk
+#####################################
+# To run report on code coverage, add option flag -DBUILD_COVERAGE=1 to cmake command
+set(BUILD_COVERAGE, 0)
```
what do you think of renaming the variable to "ENABLE_COVERAGE"?

lbk003 replied:
Done

vaasu1234 commented:
Use following CMAKE commands to check if the tools are installed
FIND_PROGRAM( GCOV_PATH gcov )
FIND_PROGRAM( LCOV_PATH lcov )
FIND_PROGRAM( GENHTML_PATH genhtml )

If not, you can abort the coverage commands gracefully. Other good techniques of adding custom target (run-coverage) like run-tests instead of POST_BUILD command.

Print out the path where the coverage files are present. 

Reference: 
https://github.com/zussel/code-coverage-test/blob/master/cmake/CodeCoverage.cmake

Commit:
Print out the path where the coverage files are present.

vaasu1234 reviewed (commented):
Also, please create an issue/story under "issues" tab and link this PR to it.

Commit:
cleaning up info files

check for GCOV, LCOV

Change BUILD_COVERAGE=>ENABLE_COVERAGE

added BUILD_TESTS dependency to ENABLE_COVERAGE

add new target for code coverage

lbk003 commented:
> Use following CMAKE commands to check if the tools are installed
> FIND_PROGRAM( GCOV_PATH gcov )
> FIND_PROGRAM( LCOV_PATH lcov )
> FIND_PROGRAM( GENHTML_PATH genhtml )
> 
> If not, you can abort the coverage commands gracefully. Other good techniques of adding custom target (run-coverage) like run-tests instead of POST_BUILD command.
> 
> Print out the path where the coverage files are present.
> 
> Reference:
> https://github.com/zussel/code-coverage-test/blob/master/cmake/CodeCoverage.cmake



> Use following CMAKE commands to check if the tools are installed
> FIND_PROGRAM( GCOV_PATH gcov )
> FIND_PROGRAM( LCOV_PATH lcov )
> FIND_PROGRAM( GENHTML_PATH genhtml )
> 
> If not, you can abort the coverage commands gracefully. Other good techniques of adding custom target (run-coverage) like run-tests instead of POST_BUILD command.
> 
> Print out the path where the coverage files are present.
> 
> Reference:
> https://github.com/zussel/code-coverage-test/blob/master/cmake/CodeCoverage.cmake

Done

Commit:
added comment for lcov version 1.13 and up

vaasu1234 commented on README.md:
````diff
@@ -104,8 +114,22 @@ You can also build individual targets using below commands
+$ make ./run-tests
+```
+
+You can run tests from the `build` folder by using commands below
````
Text needs to be updated to reflect code coverage 
- e.g., You can run tests with code coverage

vaasu1234 replied:
<img width="1423" alt="screen shot 2018-11-21 at 12 08 32 pm" src="https://user-images.githubusercontent.com/2441624/48866039-455e2b00-ed86-11e8-9bec-7317c46307f0.png">

vaasu1234 replied:
I have code coverage for c++ library files in additon to src code. Do you know why and you see the same?

lbk003 replied:
Done

vaasu1234 commented on README.md:
````diff
@@ -104,8 +114,22 @@ You can also build individual targets using below commands
+
+You can run tests from the `build` folder by using commands below
+```sh
+$ make ./run-code-coverage
````
The coverage reported is generated when I run the command `sudo make run-code-coverage`.

The command make ./run-code-coverage does nothing.  You can remove this and move the above command to this place.

lbk003 replied:
The command `make ./run-code-coverage` works on Ubuntu, but I will change it nevertheless.

lbk003 replied:
Done

vaasu1234 commented on README.md:
````diff
@@ -91,6 +94,13 @@ $ cmake .. -DBUILD_TESTS=1
 $ make
 ```
 
+If you want to view the code coverage done by the tests, run
````
Change comment to  "If you want to build tests and enable code coverage, run"

lbk003 replied:
Done

Commit:
fix the instructions for code coverage

Commit:
changing lcov exclusion path pattern

vaasu1234 commented on tests/CMakeLists.txt:
```diff
@@ -50,3 +50,51 @@ target_link_libraries(tests ${APPD_SDK_LINK_LIBS} ${CGREEN_LINK_LIBS})
+    IF(NOT GENHTML_PATH)
+        MESSAGE(FATAL_ERROR "genhtml not found! Aborting...")
+    ENDIF() # NOT GENHTML_PATH
+endif()
```
This endif() should move to the end of file to include gcov, lcov commands.

lbk003 replied:
The first block causes the cmake itself to abort, and all `make` commands will fail.

lbk003 replied:
Done

Commit:
minor edits
